Extending Governance Levels (#357)
gruetter/innerSourceTest: patterns/1-initial/governance-levels.md

> Transparent Governance Levels

Reviewer: For a discussion about alternative names for this pattern see
https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/357#issuecomment-939138250

Btw I suggest that we work out all other content of this pattern first, and the handle the name as the very last thing. Naming is so hard! :)

## Patlet

Several teams are using different InnerSource patterns and all calling it "InnerSource", so the term is not precise enough to usefully describe a set of working practices without confusion.

> Estabilishing a more accurate common language that is understood across all teams allows anyone to communicate intent or set expectations efficiently without ambiguity.

Reviewer 2: I think the part about "the term is not precise enough" misses what happens (at least where I have seen this pattern as helpful). Careful looooong explanation coming:

In Scrum, the common saying is that "either you do Scrum fully embracing all aspects of it, or you are not allowed to claim you are doing Scrum".

The thing I like about InnerSource is that the goal is different: Instead of having InnerSource projects/ teams go through a list of checkboxes and being compliant only if all checkboxes are ticked, we provide patterns to fix specific problems the organisation has. The trouble is that a lot of people have now gone through the "if you do not do all of it, you are not doing X right" training. As a result people expect a world that is either black or white: Either projects are InnerSource compliant and operate exactly as I expect them to operate, or they are not. It is tricky to teach that there is a spectrum of practices - and each place on that spectrum comes with it's own costs (in terms of shared communication) and advantages (in terms of collaboration). (The Maturity Model pattern helps to understand, which aspects are touched by InnerSource and likely should be linked here.)

Also sometimes there are corporate constraints to how far you can take InnerSource:

In some companies it may be possible to become a Trusted committer across line management chains. In others this may not be possible.

Now the funny thing is, that the same happens in Open Source (and always have been) - except the "general public" is only just now starting to understand that: https://www.sfscon.it/talks/the-economics-of-vendor-neutrality-and-vendor-domination/

My personal hope with this pattern: Teach your average engineer that there are multiple levels of InnerSource shared governance - and make them recognize that same pattern in Open Source projects:

There is a baseline (using an open source license that guarantees the four freedoms of use, study, share, improve) - but beyond that, open source projects are governed differently. At the ASF these governance specifics are defined at the project level in project bylaws. Most other projects don't make them just as transparent.

Making these specifics transparent though means that projects can evolve independently, avoiding the need to standardise everything across a huge organisation - but makes sure that the differences have to be documented locally.

What do I mean with governance differences in Open Source?

There are projects on GitHub, published purely for the pleasure of the author with no intention for long term maintenance, not intention to fix bugs submitted by users. This would be the equivalent of the governance stage "bug reports and issues welcome" - you can report the bug, but it's on me to find the time to fix it.

There are projects, where the roadmap is created in-house, hidden from public view. Where commit rights come and go with the contract of the employees of one company (think MongoDB, think Elastic, think Tensorflow etc.) - users are welcome to submit patches, they will even be mentored through. All development happens in the open, but control and strategy is never shared. This would be the equivalent of stage "contributions welcome".

There are projects that share write access, but do not share the power to decide who gets write access next. This applies to everyone who is only a committer at an Apache project. This would be the equivalent of "shared write access".

There are projects that are fully shared across multiple independent organisations - think k8s, think any Apache project - those would be shared ownership.

The same levels make sense inside of organisations.

And the same problem as with Open Source happens inside of organisations: "Look, it's on GitHub, so I can submit a PR and it will be open and developed in an open way forever" - a bit like the "if it quaks like a duck, it must a be a duck" - except not all OSS projects are the same ... https://www.slideshare.net/MikeMilinkovich/corporate-shenanigans ... helped me get a glimpse of the depth of games played.

Reviewer: Some of this text (maybe all of it?) could provide additional value in the `Rationale` section of the pattern.

Author: This is some insightful comparisons to opensource so it would be good to incorporate somehow, will take a crack at that later this week. I do stand by the "... the term is not precise enough to ..." language though because I don't think it misses the point: InnerSource is simply not a precise term and therefore needs additional qualifiers when discussing at anything other than a high level.

Author: I wanted to capture your comment here for posterity, so incorporated it into this PR in 5095397

## Story

For two projects InnerSource best practices have been adopted. Project A has a shared ownership model with [Trusted Committers](../2-structured/trusted-committer.md) from multiple teams. Project B is fully owned by one team with contributions from multiple teams. New users of either project are regularly confused about the level of influence they can gain in the respective project. This leads to long discussions, escalations and time lost on clarifications.

Project C is currently closed source and used only by team 1. Team 2 want to use project C and the leadership of the two teams negotiates options which include InnerSource practices. Agreement takes longer than expected because the "InnerSource" term did not describe a target state that could be agreed without ambiguity, and the teams had to document bespoke sub-options for their leadership to consider before a decision could be made.

## Context

> InnerSource concepts are estabilished within an organisation with multiple projects and teams adopting InnerSource. Internal InnerSource practices are not prescribed in detail and teams/projects have the autonomy to optimise for their local circumstances.

Reviewer: ```suggestion
- InnerSource concepts are established within an organisation with multiple projects and teams adopting InnerSource. 
- Internal InnerSource practices are not prescribed in detail.
- Teams/projects have the autonomy to optimise for their local circumstances.
```

Breaking this out into a list, as the checklist format makes it easier for readers to verify if this Context applies to them as well.

## Forces

- Projects adopt different InnerSource patterns and practices to optimise for their context.

> - Contributors want clarity on the level of influence they can gain in an InnerSource project they are involved with.

Reviewer: What is the motivation of the contributors here?
i.e. why do they "want clarity on the level of influence they can gain"?

Author: Good point. Changed to:

> Users want clarity on the level of influence they can gain in an InnerSource project when deciding whether or how to use it.

In my experience people are most keen on understanding what is possible with the right effort when they are making a decision on whether to use a project, or ramp up their usage or similar. In effect they are working out what they are getting into -- i.e. is this a dependency they might be blocked by or is contribution on a sensible timescale realistic.

- Leaders want clarity on InnerSource project intentions to understand the expected cost and benefits.

> - Writing a detailed description of a set of InnerSource practices requires significant effort to write and to read.

Reviewer: In terms of "makes the problem difficult", do the "write" and "read" here describe difficulties for two different scenarios or groups of people?

Author: I was simply thinking of one set of people producing some docs (write) and then the wider community consumption effort (read). But the question makes me think this is unnecessary complexity so have updated to remove `to write and to read`.

## Solution

> The solution is to create a universally understood language to describe standard building blocks for InnerSource in your organisation:

Reviewer 2: I like phrasing this as "creating a universally understood language"!

Reviewer: I am not 100% sure but this sounds like agreement to the proposed text change.
Therefore I am resolving this. Feel free to reopen if I misunderstood.

> 1. Identify the common recurring InnerSource operating models that exist in your teams and projects.

Reviewer 2: I wonder if we can identify operating models that have evolved in more than one organisation using InnerSource. Given that there are patterns in Open Source I would expect to find patterns in InnerSource orgs as well...

Reviewer: I don't quite follow.

From what I understand, you have identified 3 model below:
- Bug Reports and Issues Welcome
- Contributions Welcome
- Shared Ownership

So are you saying that you would like to know how many other orgs have used these models?

Is this just something that you are interested in in general, or would you like to add information about it to this pattern?

Reviewer: When reading this again, I think I understand it now.
I think what @MaineC meant here was whether the examples of the operating models that we are listing below are indeed common in the industry.

I am claiming that we will find this out as this pattern matures.
i.e. every time we find a new Known Instances, we can ask the respective org which models they use. That way we should find the common ones.

So for the moment I suggest that there is nothing left to do here, and I am closing this thread.

2. Document each model in detail, and give each a distinctive name.
3. Promote the use of these names to describe projects until the name's meaning is understood across the whole organisation.

Examples of common InnerSource operating models (1) are:

- **Bug Reports and Issues Welcome**: People outside the core development team may use the code. They can submit feature requests and bug reports for things they would like to see changed.
- **Contributions Welcome**: People outside the core development team may use the code. They can also make modifications and submit them to core team for inclusion.
- **Shared Ownership**: Members of different teams collaborate on the project as equal peers. Everyone has the ability to merge code. Everyone has an equal say on the project direction. Everyone has an equal say in who else to add to this group.

Examples of promoting the names (3) are:

> - Using the names within project documentation and contributing guides.

Reviewer: ```suggestion
- Using the names within project documentation and contributing guides (see also [Standard Base Documentation](../2-structured/project-setup/base-documentation.md)).
```

linking to relate pattern

- Labelling projects with the names in an [InnerSource Portal](../2-structured/innersource-portal.md).
- Presenting the names as a menu of adoption options for new InnerSource projects.
- Including the names and models in any internal InnerSource training or promotion.

## Resulting Context

- Cross team communication occurs efficiently without confusion using terms that are universally understood and centrally documented.
- Organisation leaders understand the nuances within practising InnerSource and make better informed and more precise decisions that are easier to communicate.
- Increased standardisation of InnerSource practices within the organisation as the named and documented building blocks are used by teams as a menu for adoption.
- Teams can adopt InnerSource best practices in a step-by-step way which makes adoption easier and less intimidating.

> ## Known Instances

Reviewer: @MaineC I know we talked about it but I forgot (and it is probably best to have have these conversations in the open anyways):

Are you already using this pattern at Europace? 
If yes, do you want to list your org here, with 2-3 sentences describing how you are using the pattern?

Reviewer: This comment here shouldn't block this PR. We can always add more Known Instances later.

Reviewer: No reply. Resolving.

> Flutter Entertainment define an "[Inner Source Pyramid](https://innersource.flutter.com/how/)" to describe 3 different InnerSource models: Readable Source, Guest Contributions and Maintainers in Multiple Teams. Each name is centrally documented. The use of these names is encouraged via repeated usage, direct training and categorisation of each InnerSource project.

Reviewer 2: Great to see an organisation using this pattern internally :)

> Flutter Entertainment define an "[Inner Source Pyramid](https://innersource.flutter.com/how/)" to describe 3 different InnerSource models: Readable Source, Guest Contributions and Maintainers in Multiple Teams. Each name is centrally documented. The use of these names is encouraged via repeated usage, direct training and categorisation of each InnerSource project.

Reviewer: If you like we could embed the [pyramid](https://innersource.flutter.com/how/pyramid.svg) here to illustrate your solution.

Also would it be useful to mention "Stage 0 · Closed Source" here as well, or do you consider that state not to be an InnerSource model but rather the stage below the models?

Author: We consider the 'stage 0' state not to be InnerSource, although we'd doing some work at the moment to try and expand how we express this internally to better cover other sharing models like delegating something to a team with closed source. 

Added the pyramid svg image: is the format right I copied existing pattern images `<img alt="Flutter InnerSource Pyramid" src="/assets/img/flutter-pyramid.svg" width="70%">` ..?

> Flutter Entertainment define an "[Inner Source Pyramid](https://innersource.flutter.com/how/)" to describe 3 different InnerSource models: Readable Source, Guest Contributions and Maintainers in Multiple Teams. Each name is centrally documented. The use of these names is encouraged via repeated usage, direct training and categorisation of each InnerSource project.

Reviewer: ```suggestion
Flutter Entertainment define an [InnerSource Pyramid](https://innersource.flutter.com/how/) to describe 3 different InnerSource operating models: Readable Source, Guest Contributions and Maintainers in Multiple Teams. Each name is centrally documented. The use of these names is encouraged via repeated usage, direct training and categorisation of each InnerSource project.
```

I am suggesting to use the spelling "InnerSource" here, to stay consistent when compared to other patterns. Even though I know that Flutter spells this differently.
Also we can leave the quotation marks away, as it is clear through the link that this is a unique term.

Further, should we use the exact term as described in the Solution here, i.e. "InnerSource operating models"?

## Status

> Initial

Reviewer: ```suggestion
Structured
```

With these great improvements we should absolutely move this pattern to the next level. 🚀

Reviewer: FYI I suggest that we move the pattern to the correct folder and place in the README in a subsequent PR. I can handle that once we get there.

## Story

For two projects InnerSource best practices have been adopted. Project A has a shared ownership model with [Trusted Committers](../2-structured/trusted-committer.md) from multiple teams. Project B is fully owned by one team with contributions from multiple teams. New users of either project are regularly confused about the level of influence they can gain in the respective project. This leads to long discussions, escalations and time lost on clarifications.

Project C is currently closed source and used only by team 1. Team 2 want to use project C and the leadership of the two teams negotiates options which include InnerSource practices. Agreement takes longer than expected because the "InnerSource" term did not describe a target state that could be agreed without ambiguity, and the teams had to document bespoke sub-options for their leadership to consider before a decision could be made.

## Context

> - InnerSource concepts are established within an organisation with multiple projects and teams adopting InnerSource. 

Reviewer: ```suggestion
- InnerSource concepts are established within an organisation with multiple projects and teams adopting InnerSource.
```

Reviewer: Sorry for making the linter unhappy there :)

Transparent Governance Levels

## Patlet

Several teams are using different InnerSource patterns and all calling it "InnerSource", so the term is not precise enough to usefully describe a set of working practices without confusion.

Estabilishing a more accurate common language that is understood across all teams allows anyone to communicate intent or set expectations efficiently without ambiguity.

## Problem

Several teams are using InnerSource practices. However the degree to which they welcome contributions and give equal collaboration rights to contributors differ. Despite these differences, all teams refer to their way of working as "InnerSource" without any additional qualifiers.

The result is confusion and frustration when teams collaborate as the expectation of what InnerSource means in practice is different in each team. This confusion also affects strategy planning and decisions on future InnerSource intent as the term is too vague which leads to long discussions and time lost on clarifications.

> ## Story

Reviewer: Are the two paragraphs here two different stories, both illustrating the same problem?
Or is it meant to be read as a single story?

If they are two distinct stories, then we might want to indicate that somehow.

Author: 2 examples, so I separated with sub-headings. I also incorporated [this fab comment](https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/357#discussion_r753125084) under an additional sub-heading here.

For two projects InnerSource best practices have been adopted. Project A has a shared ownership model with [Trusted Committers](../2-structured/trusted-committer.md) from multiple teams. Project B is fully owned by one team with contributions from multiple teams. New users of either project are regularly confused about the level of influence they can gain in the respective project. This leads to long discussions, escalations and time lost on clarifications.

> Project C is currently closed source and used only by team 1. Team 2 want to use project C and the leadership of the two teams negotiates options which include InnerSource practices. Agreement takes longer than expected because the "InnerSource" term did not describe a target state that could be agreed without ambiguity, and the teams had to document bespoke sub-options for their leadership to consider before a decision could be made.

Reviewer: What is meant by "sub-options" here?
i.e. sub below what?

Author: sub to "innersource", but have removed 'sub' if it is confusing.

## Context

- InnerSource concepts are established within an organisation with multiple projects and teams adopting InnerSource.
- Internal InnerSource practices are not prescribed in detail.
- Teams/projects have the autonomy to optimise for their local circumstances.

## Forces

- Projects adopt different InnerSource patterns and practices to optimise for their context.
- Users want clarity on the level of influence they can gain in an InnerSource project when deciding whether or how to use it.
- Leaders want clarity on InnerSource project intentions to understand the expected cost and benefits.
- Writing a detailed description of a set of InnerSource practices requires significant effort.

## Solution

> The solution is to create a universally understood language to describe standard building blocks for InnerSource in your organisation:

Reviewer: ```suggestion
The solution is to create a universally understood language to describe the InnerSource operating models that are used in your organisation. 

We define **InnerSource operating model** as a description of how much influence the core development team of a project ist willing to share with contributing teams. Or in other terms, the level of influence a contributing team can gain in the respective project.

The shared language for these InnerSource operating models can be established with these steps:
```

The specific building blocks that this pattern defines are the InnerSource operating models.

Therefore I am wondering if we should leave the phrase "building block" away, as that suggests that there may be even further building blocks that we want to define. Instead I would define here what we mean by "InnerSource operating models".

I tried to do that above, in parts be reusing some of the text from the left.

1. Identify the common recurring InnerSource operating models that exist in your teams and projects.
2. Document each model in detail, and give each a distinctive name.
3. Promote the use of these names to describe projects until the name's meaning is understood across the whole organisation.

Examples of common InnerSource operating models (1) are:

- **Bug Reports and Issues Welcome**: People outside the core development team may use the code. They can submit feature requests and bug reports for things they would like to see changed.
- **Contributions Welcome**: People outside the core development team may use the code. They can also make modifications and submit them to core team for inclusion.
- **Shared Ownership**: Members of different teams collaborate on the project as equal peers. Everyone has the ability to merge code. Everyone has an equal say on the project direction. Everyone has an equal say in who else to add to this group.

> Examples of promoting the names (3) are:

Reviewer: The Solution is using variations of terms when talking about the "model":
- InnerSource operating model
- model
- name

I believe that "name" is used to emphasize that we are definition a specific language here?

When reading the pattern from top to bottom, at this point I hd already forgotten what "name" was referring to. So maybe we could make it easier for the reader by calling it "model names" here, to repeat what names we are referring to?

```suggestion
Examples of promoting the model names (3) are:
```

- Using the names within project documentation and contributing guides (see also [Standard Base Documentation](../2-structured/project-setup/base-documentation.md)).
- Labelling projects with the names in an [InnerSource Portal](../2-structured/innersource-portal.md).
- Presenting the names as a menu of adoption options for new InnerSource projects.
- Including the names and models in any internal InnerSource training or promotion.

## Resulting Context

- Cross team communication occurs efficiently without confusion using terms that are universally understood and centrally documented.
- Organisation leaders understand the nuances within practising InnerSource and make better informed and more precise decisions that are easier to communicate.
- Increased standardisation of InnerSource practices within the organisation as the named and documented building blocks are used by teams as a menu for adoption.
- Teams can adopt InnerSource best practices in a step-by-step way which makes adoption easier and less intimidating.

## Known Instances

> <img alt="Flutter InnerSource Pyramid" src="/assets/img/flutter-pyramid.svg" width="70%">

Reviewer: Note: We need to check whether SVG gets displayed correctly when publishing to our book (on gitbook.com). If not, then we can use PNG instead.

Reviewer: Note to self: We will also need markdown notation for the image, in order for gitbook to pick it up correctly.

Reviewer: ```suggestion
![InnerSource Pyramid used by Flutter Entertainment](../../assets/img/flutter-pyramid.svg)
```
